doc/site/docs/02-core-concepts/07-middleware.md

```diff
@@ -0,0 +1,292 @@
+
+They work by wrapping handlers with additional functionality. Think of it as a series of layers around your core handler, each layer can inspect and modify requests before they reach your handler, and responses after they leave your handler.
+
+## What is Middleware?
+
+A `Middleware` is a function that takes a handler and returns a new handler with enhanced functionality. This pattern allows you to compose multiple pieces of functionality together, creating a processing pipeline where each middleware layer adds its own behavior.
+
+```dart
+typedef Middleware = Handler Function(Handler innerHandler);
+```
+
+### Handler Wrapping
+
+Every middleware function receives an inner handler and returns a new handler that wraps the original. This wrapper can:
+
+- **Inspect requests** before passing them to the inner handler
+- **Modify requests** by creating new request objects
+- **Short-circuit processing** by returning a response without calling the inner handler
+- **Transform responses** after the inner handler completes
+- **Handle errors** that occur in the inner handler
+
+Here's a simple example:
+
+```dart title="middleware_example.dart"
+import 'package:relic/relic.dart';
+
+// A middleware that adds a custom header
+Middleware addHeaderMiddleware() {
+  return (Handler innerHandler) {
+    return (NewContext ctx) async {
+      // Call the inner handler
+      final result = await innerHandler(ctx);
+      
+      // Add custom header to response
+      if (result is ResponseContext) {
+        final newResponse = result.response.copyWith(
+          headers: result.response.headers.transform(
+            (mh) => mh['X-Custom-Header'] = ['Hello from middleware!'],
+          ),
+        );
+        return result.respond(newResponse);
+      }
+      
+      return result;
+    };
+  };
+}
+```
+
+## Using Middlewares with Router
+
+### `router.use()` for Applying Middleware
+
+Relic's `RelicRouter` provides a convenient `use()` method for applying middleware to specific path patterns. This is the preferred way to add middleware in modern Relic applications.
+
+```dart
+import 'package:relic/relic.dart';
+
+final router = RelicRouter()
+  // Apply logging to all routes
+  ..use('/', logRequests())
+  
+  // Apply authentication to API routes
+  ..use('/api', authMiddleware())
+  
+  // Define your routes
+  ..get('/api/users', usersHandler)
+  ..post('/api/users', createUserHandler);
+```
+
+### Global vs Route-Specific Middlewares
+
+There are two types of categories of middlewares: global and route-specific. Global middlewares apply to all routes in your application, while route-specific middlewares apply only to routes under a specific path.
+
+**Global Middleware:** applies to all routes in your application:
+
+```dart
+final router = RelicRouter()
+  // Global middleware - applies to ALL routes
+  ..use('/', logRequests())
+  ..use('/', corsMiddleware())
```

Comment:
Hmm.. these are actually also route specific, they just happens to apply for all routes below `/` which means everything that is actually matched.

But if a route is not matched, then these will no be used on the `fallback`.

Truly global middleware, that is run on every request, even those that end in fallback would have to be setup either with a pipeline.

I think we should add `useBefore` to setup truly global middleware 🤔 .. or just show people how to use a pipeline for that.